File: data_analysis.py
from pyspark.sql import SparkSession
import shutil
import os

def create_csv(df,category,topic,indicator):
    base_dir = "reports"
    category_dir = os.path.join(base_dir, category)
    topic_dir = os.path.join(category_dir, topic)
    
    os.makedirs(topic_dir, exist_ok=True)

    modified_indicator = indicator.replace(".","-")

    namefile = os.path.join(topic_dir, f"{category}_{topic}_{modified_indicator}.csv")

    temp_dir = os.path.join(base_dir, "temp_WDI_" + topic)

    df.coalesce(1).write.option("header", "true").csv(temp_dir)

    temp_file = [file for file in os.listdir(temp_dir) if file.startswith("part-")][0]

    shutil.move(os.path.join(temp_dir, temp_file), namefile)

    shutil.rmtree(temp_dir)

spark = SparkSession.builder \
    .appName("WDI Analysis") \
    .getOrCreate()

file_path = "data/Countries_Indicators.csv"
df = spark.read.csv(file_path, header=True, inferSchema=True)

# Country Categories
innovative_countries = ["ESP", "USA", "CHE", "SWE", "GBR", "NLD"]
neighbour_countries = ["ESP","PRT","FRA","ITA","MAR","DZA"]
countries=[innovative_countries, neighbour_countries]

# Indicators by topic
economic_indicators = ["NY.GDP.MKTP.CD","NY.GDP.PCAP.CD","NY.GDP.MKTP.KD.ZG","NY.GNP.PCAP.CD","SI.POV.GINI","FP.CPI.TOTL.ZG","IC.BUS.EASE.XQ"]
education_indicators = ["SE.XPD.TOTL.GD.ZS","SE.SEC.ENRR","SE.TER.ENRR","SE.TER.CUAT.BA.ZS"]
energy_indicators = ["EG.USE.PCAP.KG.OE","EG.ELC.ACCS.ZS"]
research_dev_indicators = ["GB.XPD.RSDV.GD.ZS","SP.POP.SCIE.RD.P6","SP.POP.TECH.RD.P6","IP.PAT.RESD","IP.PAT.NRES","IP.JRN.ARTC.SC","IT.NET.USER.ZS","IT.CEL.SETS.P2","IT.NET.BBND.P2","IT.NET.SECR.P6","TX.VAL.TECH.MF.ZS","TX.VAL.TECH.CD"]
indicators=[economic_indicators, education_indicators, energy_indicators, research_dev_indicators]

#Columns select
years = ['2001', '2002', '2003', '2004', '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021', '2022', '2023']
fixed_columns = ["Country Name", "Country Code", "Indicator Name"]
selected_columns = fixed_columns + years

categories=["innovative","neighbour"]
topics=["economic","education","energy","research_dev"]

for i, category in enumerate(categories):
    for j, topic in enumerate(topics):
        for indicator in indicators[j]:
            df_filter = df.filter((df["Country Code"].isin(countries[i])) & (df["Indicator Code"] == indicator))
            df_select = df_filter.select(*selected_columns)

            create_csv(df_select,category,topic,indicator)

File: src/main/scala/DataAnalysis.scala
import org.apache.spark.sql.{DataFrame, SparkSession}
import java.io.File
import java.nio.file.{Files, Paths, StandardCopyOption}

object DataAnalysis {
  def createCSV(df: DataFrame, category: String, topic: String, indicator: String): Unit = {
    val baseDir = "reports"
    val categoryDir = Paths.get(baseDir, category).toString
    val topicDir = Paths.get(categoryDir, topic).toString

    new File(topicDir).mkdirs()

    val modifiedIndicator = indicator.replace(".", "-")
    val nameFile = Paths.get(topicDir, s"${category}_${topic}_${modifiedIndicator}.csv").toString

    val tempDir = Paths.get(baseDir, s"temp_WDI_$topic").toString

    df.coalesce(1).write.option("header", "true").csv(tempDir)

    val tempFile = new File(tempDir).listFiles().filter(_.getName.startsWith("part-")).head

    Files.move(tempFile.toPath, Paths.get(nameFile), StandardCopyOption.REPLACE_EXISTING)

    new File(tempDir).listFiles().foreach(_.delete())
    new File(tempDir).delete()
  }

  def main(args: Array[String]): Unit = {
    val spark = SparkSession.builder
      .appName("WDI Analysis")
      .config("spark.master", "local[*]")
      .getOrCreate()

    val filePath = "data/Countries_Indicators.csv"
    val df = spark.read.option("header", "true").option("inferSchema", "true").csv(filePath)

    // Country Categories
    val innovativeCountries = List("ESP", "USA", "CHE", "SWE", "GBR", "NLD")
    val neighbourCountries = List("ESP", "PRT", "FRA", "ITA", "MAR", "DZA")
    val countries = List(innovativeCountries, neighbourCountries)

    // Indicators by topic
    val economicIndicators = List("NY.GDP.MKTP.CD", "NY.GDP.PCAP.CD", "NY.GDP.MKTP.KD.ZG", "NY.GNP.PCAP.CD", "SI.POV.GINI", "FP.CPI.TOTL.ZG", "IC.BUS.EASE.XQ")
    val educationIndicators = List("SE.XPD.TOTL.GD.ZS", "SE.SEC.ENRR", "SE.TER.ENRR", "SE.TER.CUAT.BA.ZS")
    val energyIndicators = List("EG.USE.PCAP.KG.OE", "EG.ELC.ACCS.ZS")
    val researchDevIndicators = List("GB.XPD.RSDV.GD.ZS", "SP.POP.SCIE.RD.P6", "SP.POP.TECH.RD.P6", "IP.PAT.RESD", "IP.PAT.NRES", "IP.JRN.ARTC.SC", "IT.NET.USER.ZS", "IT.CEL.SETS.P2", "IT.NET.BBND.P2", "IT.NET.SECR.P6", "TX.VAL.TECH.MF.ZS", "TX.VAL.TECH.CD")
    val indicators = List(economicIndicators, educationIndicators, energyIndicators, researchDevIndicators)

    // Columns to select
    val years = (2001 to 2023).map(_.toString).toList
    val fixedColumns = List("Country Name", "Country Code", "Indicator Name")
    val selectedColumns = fixedColumns ++ years

    val categories = List("innovative", "neighbour")
    val topics = List("economic", "education", "energy", "research_dev")

    for ((category, i) <- categories.zipWithIndex) {
      for ((topic, j) <- topics.zipWithIndex) {
        for (indicator <- indicators(j)) {
          val dfFilter = df.filter(df("Country Code").isin(countries(i): _*) && df("Indicator Code") === indicator)
          val dfSelect = dfFilter.select(selectedColumns.head, selectedColumns.tail: _*)

          createCSV(dfSelect, category, topic, indicator)
        }
      }
    }
    spark.stop()
  }
}
